web/packages/shared/components/UnifiedResources/UnifiedResources.tsx

```diff
@@ -267,21 +288,39 @@ export function UnifiedResources(props: UnifiedResourcesProps) {
     updateUnifiedResourcesPreferences({ defaultTab: value });
   };
 
-  const $pinAllButton = (
-    <ButtonBorder
-      onClick={() => handlePinSelected(shouldUnpin)}
-      textTransform="none"
-      disabled={pinning.kind === 'not-supported'}
-      css={`
-        border: none;
-        color: ${props => props.theme.colors.brand};
-      `}
-    >
-      <PushPin color="brand" size={16} mr={2} />
-      {shouldUnpin ? 'Unpin ' : 'Pin '}
-      Selected
-    </ButtonBorder>
-  );
+  const getSelectedResources = () => {
+    return resources
+      .filter(({ resource }) =>
+        selectedResources.includes(generateUnifiedResourceKey(resource))
+      )
+      .map(({ resource }) => ({
+        resource: resource,
+        unifiedResourceId: generateUnifiedResourceKey(resource),
+      }));
+  };
+
+  const bulkActionsAndPinning = () => {
+    if (pinning.kind === 'hidden') {
+      return bulkActions;
+    }
+
+    return [
+      ...bulkActions,
+      {
+        key: 'pin_resource',
+        text: shouldUnpin ? 'Unpin Selected' : 'Pin Selected',
+        Icon: PushPin,
+        tooltip:
+          pinning.kind !== 'not-supported'
+            ? PINNING_NOT_SUPPORTED_MESSAGE
+            : null,
+        disabled:
+          pinning.kind === 'not-supported' ||
+          updatePinnedResourcesAttempt.status === 'processing',
+        action: () => handlePinSelected(shouldUnpin),
+      } as BulkActionProps,
+    ];
+  };
```

Comment:
This is really a nit, but IMO if we can just use a field, I'd do it:
```suggestion
  const bulkActionsAndPinning: BulkActionProps[] = [
    ...bulkActions,
    pinning.kind !== 'hidden' && {
      key: 'pin_resource',
      text: shouldUnpin ? 'Unpin Selected' : 'Pin Selected',
      Icon: PushPin,
      tooltip:
        pinning.kind !== 'not-supported' ? PINNING_NOT_SUPPORTED_MESSAGE : null,
      disabled:
        pinning.kind === 'not-supported' ||
        updatePinnedResourcesAttempt.status === 'processing',
      action: () => handlePinSelected(shouldUnpin),
    },
  ].filter(Boolean);
```

Also, it's better to avoid `as` - it can hide some type errors.

Reply:
Ah, the actions are rendered conditionally, so I think it makes sense to return them from a function. I'd only use a return type instead of `as BulkActionProps`.